core/math/delaunay_2d.h

```diff
@@ -48,117 +49,114 @@ class Delaunay2D {
-				(ab * (p3.x - p2.x) + cd * (p1.x - p3.x) + ef * (p2.x - p1.x)) / (p1.y * (p3.x - p2.x) + p2.y * (p1.x - p3.x) + p3.y * (p2.x - p1.x)));
+		Vector2 O = (b * a.length_squared() - a * b.length_squared()) / (a.cross(b) * 2.0f);
 
-		circum *= 0.5;
-		float r = p1.distance_squared_to(circum);
-		float d = p_vertices[p_vertex].distance_squared_to(circum);
-		return d <= r;
-	}
+		// Rotate by 90 degrees.
+		real_t temp = O.y;
+		O.y = -O.x;
+		O.x = temp;
 
-	static bool edge_compare(const Vector<Vector2> &p_vertices, const Edge &p_a, const Edge &p_b) {
-		if (p_vertices[p_a.edge[0]].is_equal_approx(p_vertices[p_b.edge[0]]) && p_vertices[p_a.edge[1]].is_equal_approx(p_vertices[p_b.edge[1]])) {
-			return true;
-		}
-
-		if (p_vertices[p_a.edge[0]].is_equal_approx(p_vertices[p_b.edge[1]]) && p_vertices[p_a.edge[1]].is_equal_approx(p_vertices[p_b.edge[0]])) {
-			return true;
-		}
+		triangle.circum_radius_squared = O.length_squared();
+		triangle.circum_center = O + p_vertices[p_a];
 
-		return false;
+		return triangle;
 	}
 
 	static Vector<Triangle> triangulate(const Vector<Vector2> &p_points) {
 		Vector<Vector2> points = p_points;
 		Vector<Triangle> triangles;
 
-		Rect2 rect;
-		for (int i = 0; i < p_points.size(); i++) {
-			if (i == 0) {
-				rect.position = p_points[i];
-			} else {
-				rect.expand_to(p_points[i]);
-			}
+		// Return an empty vector immediately for less than 2 points.
+		int point_count = p_points.size();
+		if (point_count < 2) {
+			return triangles;
+		}
+
+		// Get a bounding rectangle.
+		Rect2 rect = Rect2(p_points[0], Size2());
+		for (int i = 1; i < point_count; i++) {
+			rect.expand_to(p_points[i]);
 		}
 
-		float delta_max = MAX(rect.size.width, rect.size.height);
+		real_t delta_max = MAX(rect.size.width, rect.size.height);
 		Vector2 center = rect.get_center();
 
-		points.push_back(Vector2(center.x - 20 * delta_max, center.y - delta_max));
-		points.push_back(Vector2(center.x, center.y + 20 * delta_max));
-		points.push_back(Vector2(center.x + 20 * delta_max, center.y - delta_max));
+		// Construct a bounding triangle around the rectangle.
+		points.push_back(Vector2(center.x - delta_max * 16, center.y - delta_max));
+		points.push_back(Vector2(center.x, center.y + delta_max * 16));
+		points.push_back(Vector2(center.x + delta_max * 16, center.y - delta_max));
 
-		triangles.push_back(Triangle(p_points.size() + 0, p_points.size() + 1, p_points.size() + 2));
+		Triangle bounding_triangle = create_triangle(points, point_count + 0, point_count + 1, point_count + 2);
+		triangles.push_back(bounding_triangle);
 
-		for (int i = 0; i < p_points.size(); i++) {
+		for (int i = 0; i < point_count; i++) {
 			Vector<Edge> polygon;
 
+			// Save the edges of the triangles whose circumcircles contain the i-th vertex. Delete the triangles themselves.
 			for (int j = 0; j < triangles.size(); j++) {
```

Comment:
Note the triangle at index `j` can be removed (using `remove_at` down below), meaning in such case all the triangles within the array after the removed one will need to be shifted by 1. Thus it's more efficient to do it in reverse (as in #65280), less overall shifting. Also currently in this PR (with iterating in non-reversed manner) there's a bug as when removing a triangle one triangle would be skipped as in such case the `j` index is still incremented although the triangles after the removed one will end up at their previous index minus one (because of the removal/shifting).
```suggestion
			for (int j = triangles.size() - 1; j >= 0; j--) {
```

---

This whole loop could be potentially further improved by not using `remove_at` at all and manually placing each not removed triangle directly at the correct final index when iterating over. And resizing once at the end. Something like:

```suggestion
			int triangles_left_count = 0;
			Triangle *triangles_ptrw = triangles.ptrw();
			for (int j = 0; j < triangles.size(); j++) {
				if (points[i].distance_squared_to(triangles[j].circum_center) < triangles[j].circum_radius_squared) {
					polygon.push_back(Edge(triangles[j].points[0], triangles[j].points[1]));
					polygon.push_back(Edge(triangles[j].points[1], triangles[j].points[2]));
					polygon.push_back(Edge(triangles[j].points[2], triangles[j].points[0]));
				} else {
					//if (triangles_left_count != j) { // Not sure about this condition, might end up being faster without it, would need benchmarking.
					triangles_ptrw[triangles_left_count] = triangles[j];
					//}
					triangles_left_count++;
				}
			}
			triangles.resize(triangles_left_count);
```

Reply:
I tested this and it didn't improve the performance.